### PR TITLE
Runner proxy

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

File: runner-proxy/server/handle.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		// Set the url scheme to http
+		r.URL.Scheme = "http"
+
+		helpers.Logger.LogDebug(helpers.GetRequestID(ctx), fmt.Sprintf("Proxy is making request to host (%s) port (%s)", ogHost, ogPort), nil)
+
+		// Update the ttl of cached deployment
+		id := fmt.Sprintf("%s-%s-%s", project, service, ogVersion)
+		exist := s.cache.GetDeployment(id)
+		if exist {
+			_ = helpers.Response.SendOkayResponse(ctx, http.StatusOK, w)
```
Comment (YourTechBud) [inline runner-proxy/server/handle.go:54]: You are supposed to proxy the request here

File: runner-proxy/server/handle.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			return
+		}
+
+		// get token from header
+		token := utils.GetToken(r)
```
Comment (YourTechBud) [inline runner-proxy/server/handle.go:59]: You will have to generate a token. The request will not have runner's token

File: runner-proxy/server/handle.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		var vPtr interface{}
+		url := fmt.Sprintf("/v1/runner/%s/scale-up/%s/%s", project, service, ogVersion)
+		if err := utils.MakeHTTPRequest(ctx, "POST", url, token, "", map[string]interface{}{}, &vPtr); err != nil {
+			_ = helpers.Response.SendErrorResponse(ctx, w, http.StatusServiceUnavailable, err)
+			return
+		}
+
+		// Wait for the service to scale up
+		url = fmt.Sprintf("/v1/runner/%s/wait/%s/%s", project, service, ogVersion)
+		if err := s.debounce.Wait(fmt.Sprintf("proxy-%s-%s-%s", project, service, ogVersion), func() error {
```
Comment (YourTechBud) [inline runner-proxy/server/handle.go:71]: We discussed that we don't need debounce logic

File: runner-proxy/utils/cache.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+// GetDeployment check if the element exist in the TTLMap
+func (m *TTLMap) GetDeployment(k string) bool {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	it, ok := m.m[k]
+	if ok {
+		if time.Now().Unix()-it.lastAccess < int64(120) {
+			m.m[k] = &Item{lastAccess: time.Now().Unix()}
```
Comment (YourTechBud) [inline runner-proxy/utils/cache.go:62]: This isn't required since you are calling put anyways

File: runner-proxy/utils/cache.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		m.m[k] = it
+	} else {
+		m.m[k] = &Item{lastAccess: time.Now().Unix()}
+	}
+}
+
+// GetDeployment check if the element exist in the TTLMap
+func (m *TTLMap) GetDeployment(k string) bool {
+	m.l.Lock()
+	defer m.l.Unlock()
```
Comment (YourTechBud) [inline runner-proxy/utils/cache.go:57]: Can be read lock

File: runner-proxy/utils/http.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		return err
+	}
+
+	// Add the headers
+	if token != "" {
+		// Add the token only if its provided
+		req.Header.Add("Authorization", "Bearer "+token)
+	}
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("x-sc-token", "Bearer "+scToken)
```
Comment (YourTechBud) [inline runner-proxy/utils/http.go:45]: Don't need this since it's an internal request

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Has this been thoroughly tested?

**Comment by prithvi081099** [comment]
> Has this been thoroughly tested?

No, I am still testing the code. Facing some issues trying to solve them.

**Pull request closed without merging**